README
== About ==

Maps is the MediaWiki extension that provides the ability to visualize geographic data
with dynamic, JavaScript based, mapping API's such as Google Maps and OpenLayers in your
wiki pages. It has build in support for geocoding, displaying maps, displaying markers,
adding pop-ups, and more. Maps allows extensive customization of your maps, either per
map, or via a general setting affecting all maps on your wiki.

Due to Maps modular build, modifying the mapping service of a map is as easy as changing
a single map property! These mapping services include Google Maps, Yahoo! Maps, OpenLayers
and OpenStreetMap. These also allow you to display maps with Google Earth, OpenStreetMaps,
Bing maps and others.

Notes on installing Maps are found in the file INSTALL.


== Contributing ==

If you have bug reports or requests, please add them to the Talk page [0]. You can also
send them to Jeroen De Dauw, jeroendedauw -at- gmail.com, and Yaron Koren, at yaron57 -at-
gmail.com.

[0] http://www.mediawiki.org/w/index.php?title=Extension_talk:Maps

For more info, see http://mapping.referata.com/wiki/Mapping_on_MediaWiki

== Credits to other projects ==

=== jQuery ===

This extension uses code from the jQuery library.
jQuery is dual licensed under the MIT [0] and GPL [1] licenses.

=== OpenLayers ===

This extension includes code from the OpenLayers application.
OpenLayers is an open-source product released under a BSD-style license [2].

=== geoxml3 ===

This extension includes a copy of the geoxml3 KML processor.
geoxml3 is released under the Apache License 2.0 license [3].

=== google-maps-utility-library-v3 ===

This extension icnludes code from the google-maps-utility-library-v3 (googleearth.js).
It is released under the Apache License 2.0 license [3].

=== OpenStreetMap.js ===

This extension includes the OpenStreetMap.js file which can be found here:
http://www.openstreetmap.org/openlayers/OpenStreetMap.js

[0] http://www.opensource.org/licenses/mit-license.php
[1] http://www.opensource.org/licenses/gpl-license.php
[2] http://svn.openlayers.org/trunk/openlayers/license.txt
[3] http://www.apache.org/licenses/LICENSE-2.0

RELEASE-NOTES
These are the release notes for the Maps extension.

Extension page on mediawiki.org: https://www.mediawiki.org/wiki/Extension:Maps
Latest version of the release notes: https://gerrit.wikimedia.org/r/gitweb?p=mediawiki/extensions/Maps.git;a=blob;f=RELEASE-NOTES


=== Maps 3.0 ===
(dev)

In this version a big part of the PHP codebase has been rewritten to decrease technical debt and thus
facilatete maintenance, new feature delopment and debugging. Many tests have been added and a lot of
bugs have been found and fixed. No big new features have been added.

; Compatibility changes

* Changed minimum Validator version from 0.5 to 1.0.
* DataValues is now required. Version 0.1 or later.
* ValueParsers is now required. Version 0.1 or later.
* ValueFormatters is now required. Version 0.1 or later.

; New features

* Added leaflet service

; Breaking internal changes

* Moved classes into Maps namespace
* Removed all Criteria classes
* Removed all Manipulation classes
* Removed MapsCoordinateParser
* Geocoding interfaces changed
* MapsLocation interface changed

; Bug fixes

* Fixed autoinfowindows functionality.
* Fixed various bugs in geocoordinate parsing and formatting.

=== Maps 2.0 ===
(2012-10-05)

; Compatibility changes

* Changed minimum PHP version from 5.2 to 5.3.
* Changed minimum MediaWiki version from 1.17 to 1.18.
* Changed minimum Validator version from 0.4 to 0.5.
* Removed support for the deprecated Google Maps v2 API.
* Removed support for the now unsupported Yahoo! Maps API and associated geocoding service.
* Temporary disabled OSM service (you can still use OSM with the OpenLayers service).

; New features

* Added support for defining an inline label to markers to GoogleMaps.
* Added support for marker clustering to Google Maps.
* Added support for grouping locations.
* Added support for defining image overlays (ground overlays) in Google Maps.
* Added support for defining lines, polygons, rectangles and circles using wikitext for Google Maps and OpenLayers.
* Added a graphical map editing tool that allows exporting to and importing from simple wikitext (Google Maps only).
* Added "copycoords" parameter to Google Maps and OpenLayers that allows copying coordinates after right clicking a location on a map.
* Added "minzoom" and "maxzoom" parameters to #display_map.
* Added support for using the Google JS API key with Google Maps (for increased map display and geocoding call limits).
* Added support for searching markers (searchmarkers=all/title) in Google Maps and OpenLayers
* Added support for creating static maps in OpenLayers and GoogleMaps (static=on)
* Added positional parameter to show polygons only on hover.
* Added an optional link parameter as an alternative to popup bubble with text and title
* Added an optional visitedicon parameter (both global and marker parameter), that will change the icon of a marker on click.

; Other improvements

* Merged display_map and display_point(s) into a single parser function: display_map (display_points is now an alias).
* Updates parameter definitions from Validator 0.4.x to Validator 0.5.x.
* Improved script loading.
* Added various unit tests that caught some bugs and will now prevent regressions.

; Bug fixes

* Fixed JavaScript error on some special pages due to incorrect order of map initialization.
* Fixed partially broken kml functionality.

=== Maps 1.0.5 ===
(2011-11-30)

* Fixed display of attribution control for OpenLayers.
* Fixed to big precision of geographic coordinates in decimal minutes format (bug 32407).

=== Maps 1.0.4 ===
(2011-10-15)

* Updated OpenLayers from 2.10 to 2.11.
* Fixed bug in adding adittional markers for Google Maps v3 (mainly affecting the Semantic Maps form input).

=== Maps 1.0.3 ===
(2011-09-14)

* Added API module for geocoding.
* Added 'geocoding' right.
* Added kmlrezoom parameter for Google Maps v3 and general $egMapsRezoomForKML setting.
* Fixed Google Maps v3 JavaScript issue occuring on MediaWiki 1.17.

=== Maps 1.0.2 ===
(2011-08-24)

* Fixed Google Maps v3 JavaScript issue occuring when using Google Earth on unsupported systems.
* Fixed internationalization of distances (bug 30467).

=== Maps 1.0.1 ===
(2011-08-17)

* Added language parameter to the mapsdoc hook.

* Use of Validator 0.4.10s setMessage method instead of setDescription for better i18n.

* Fixed zoom and types parameters for Google Maps v3.

* Minor improvement to script loading.

* Added support for Google Earth in Google Maps v3.

* Added tilt parameter for Google Earth in Google Maps v3.

=== Maps 1.0 ===
(2011-07-19)

This version branched from Maps 0.7.x at version 0.7.3.

==== New features ====

* Added full Google Maps v3 support and set it as the default mapping service.

* Added new geocoder making use of the new GeoNames API.

* Added support for the auto-documentation features for parser hooks introduced in Validator 0.4.3.

* Added resizeable parameter to all mapping services except OSM.

==== Removed features ====

* Removed compatibility with pre MediaWiki 1.17.

* Removed overlays parameter for Google Maps v2.

* Removed the previously deprecated "display map", "display point" and "display points" parser hooks. Use their underscored equivalents, ie "display_map".

==== Internal improvements ====

* Usage of the Resource Loader for all scripts and stylesheets.

* Rewrote all the map JavaScript to jQuery plugins.

* Rewrote the way parameters are translated to JavaScript. Now one big PHP object is json_encoded.

* Improved KML formatter.

* Use of Google Maps geocoding service v3 instead of v2.

* Completed coordinate and distance parser/formatter unit tests and made them compliant with the MediaWiki unit testing support.

==== Bug fixes ====

* Fixed geocoding service overriding based on mapping service (merged in from Maps 0.7.5).

* Fixed fatal error occurring when running maintenance/refreshLinks.php.

* Fixed DMS coordinate parsing issue (bug 29419).


Maps.php
<?php

/**
 * Initialization file for the Maps extension.
 * https://github.com/JeroenDeDauw/Maps
 *
 * @licence GNU GPL v2+
 * @author Jeroen De Dauw < jeroendedauw@gmail.com >
 */

if ( !defined( 'MEDIAWIKI' ) ) {
	die( 'Not an entry point.' );
}

if ( defined( 'Maps_VERSION' ) ) {
	// Do not initialize more than once.
	return 1;
}

define( 'Maps_VERSION' , '3.0' );

// Include the composer autoloader if it is present.
if ( is_readable( __DIR__ . '/vendor/autoload.php' ) ) {
	include_once( __DIR__ . '/vendor/autoload.php' );
}

// Only initialize the extension when all dependencies are present.
if ( !defined( 'Validator_VERSION' ) ) {
	throw new Exception( 'You need to have Validator installed in order to use Maps' );
}

if ( version_compare( $GLOBALS['wgVersion'], '1.18c' , '<' ) ) {
	throw new Exception( 'This version of Maps requires MediaWiki 1.18 or above; use Maps 1.0.x for MediaWiki 1.17 and Maps 0.7.x for older versions.' );
}

call_user_func( function() {
	global $wgExtensionCredits;
	global $wgResourceModules, $wgGroupPermissions, $egMapsNamespaceIndex, $wgStyleVersion;
	global $egMapsStyleVersion, $wgHooks, $wgExtensionMessagesFiles;

	$wgExtensionCredits['parserhook'][] = array(
		'path' => __FILE__ ,
		'name' => 'Maps' ,
		'version' => Maps_VERSION ,
		'author' => array(
			'[http://www.mediawiki.org/wiki/User:Jeroen_De_Dauw Jeroen De Dauw]'
		) ,
		'url' => 'https://github.com/JeroenDeDauw/Maps/blob/master/README.md#maps' ,
		'descriptionmsg' => 'maps-desc'
	);

	// The different coordinate notations.
	define( 'Maps_COORDS_FLOAT' , 'float' );
	define( 'Maps_COORDS_DMS' , 'dms' );
	define( 'Maps_COORDS_DM' , 'dm' );
	define( 'Maps_COORDS_DD' , 'dd' );

	$egMapsDir = __DIR__ . '/';

	$egMapsStyleVersion = $wgStyleVersion . '-' . Maps_VERSION;

	$wgExtensionMessagesFiles['Maps'] 				= __DIR__ . '/Maps.i18n.php';
	$wgExtensionMessagesFiles['MapsMagic'] 			= __DIR__ . '/Maps.i18n.magic.php';
	$wgExtensionMessagesFiles['MapsNamespaces'] 	= __DIR__ . '/Maps.i18n.namespaces.php';
	$wgExtensionMessagesFiles['MapsAlias'] 			= __DIR__ . '/Maps.i18n.alias.php';

	$wgResourceModules = array_merge( $wgResourceModules, include 'Maps.resources.php' );

	$wgAPIModules['geocode'] = 'Maps\Api\Geocode';

	// Register the initialization function of Maps.
	$GLOBALS['wgExtensionFunctions'][] = function () {
		wfRunHooks( 'MappingServiceLoad' );
		wfRunHooks( 'MappingFeatureLoad' );

		if ( in_array( 'googlemaps3', $GLOBALS['egMapsAvailableServices'] ) ) {
			global $wgSpecialPages, $wgSpecialPageGroups;

			$wgSpecialPages['MapEditor'] = 'SpecialMapEditor';
			$wgSpecialPageGroups['MapEditor'] = 'maps';
		}

		return true;
	};

	$wgHooks['AdminLinks'][]                = 'MapsHooks::addToAdminLinks';
	$wgHooks['ArticleFromTitle'][]          = 'MapsHooks::onArticleFromTitle';
	$wgHooks['MakeGlobalVariablesScript'][] = 'MapsHooks::onMakeGlobalVariablesScript';
	$wgHooks['CanonicalNamespaces'][]       = 'MapsHooks::onCanonicalNamespaces';	$wgHooks['LoadExtensionSchemaUpdates'][] = 'MapsHooks::onLoadExtensionSchemaUpdates';
	$wgHooks['ArticlePurge'][]              = 'MapsHooks::onArticlePurge';
	$wgHooks['LinksUpdateConstructed'][]    = 'MapsHooks::onLinksUpdateConstructed';
	$wgHooks['ParserAfterTidy'][]           = 'MapsHooks::onParserAfterTidy';
	$wgHooks['ParserClearState'][]          = 'MapsHooks::onParserClearState';

	// Parser hooks

	// Required for #coordinates.
	$wgHooks['ParserFirstCallInit'][] = function( Parser &$parser ) {
		$instance = new MapsCoordinates();
		return $instance->init( $parser );
	};

	$wgHooks['ParserFirstCallInit'][] = function( Parser &$parser ) {
		$instance = new MapsDisplayMap();
		return $instance->init( $parser );
	};

	$wgHooks['ParserFirstCallInit'][] = function( Parser &$parser ) {
		$instance = new MapsDistance();
		return $instance->init( $parser );
	};

	$wgHooks['ParserFirstCallInit'][] = function( Parser &$parser ) {
		$instance = new MapsFinddestination();
		return $instance->init( $parser );
	};

	$wgHooks['ParserFirstCallInit'][] = function( Parser &$parser ) {
		$instance = new MapsGeocode();
		return $instance->init( $parser );
	};

	$wgHooks['ParserFirstCallInit'][] = function( Parser &$parser ) {
		$instance = new MapsGeodistance();
		return $instance->init( $parser );
	};

	$wgHooks['ParserFirstCallInit'][] = function( Parser &$parser ) {
		$instance = new MapsMapsDoc();
		return $instance->init( $parser );
	};

	$wgHooks['ParserFirstCallInit'][] = function( Parser &$parser ) {
		$instance = new MapsLayerDefinition();
		return $instance->init( $parser );
	};

	// Geocoders

	// Registration of the GeoNames service geocoder.
	$wgHooks['GeocoderFirstCallInit'][] = 'MapsGeonamesGeocoder::register';

	// Registration of the Google Geocoding (v2) service geocoder.
	$wgHooks['GeocoderFirstCallInit'][] = 'MapsGoogleGeocoder::register';

	// Registration of the geocoder.us service geocoder.
	$wgHooks['GeocoderFirstCallInit'][] = 'MapsGeocoderusGeocoder::register';

	// Layers

	// Registration of the image layer type.
	$wgHooks['MappingLayersInitialization'][] = 'MapsImageLayer::register';

	// Mapping services

	// Include the mapping services that should be loaded into Maps.
	// Commenting or removing a mapping service will make Maps completely ignore it, and so improve performance.

	// Google Maps API v3
	// TODO: improve loading mechanism
	include_once $egMapsDir . 'includes/services/GoogleMaps3/GoogleMaps3.php';

	// OpenLayers API
	// TODO: improve loading mechanism
	include_once $egMapsDir . 'includes/services/OpenLayers/OpenLayers.php';

	// Leaflet API
	// TODO: improve loading mechanism
	include_once $egMapsDir . 'includes/services/Leaflet/Leaflet.php';


	require_once __DIR__ . '/Maps_Settings.php';

	define( 'Maps_NS_LAYER' , $egMapsNamespaceIndex + 0 );
	define( 'Maps_NS_LAYER_TALK' , $egMapsNamespaceIndex + 1 );

	$wgAvailableRights[] = 'geocode';

	// Users that can geocode. By default the same as those that can edit.
	foreach ( $wgGroupPermissions as $group => $rights ) {
		if ( array_key_exists( 'edit' , $rights ) ) {
			$wgGroupPermissions[$group]['geocode'] = $wgGroupPermissions[$group]['edit'];
		}
	}

	global $wgParamDefinitions;

	$wgParamDefinitions['mappingservice'] = array(
		'definition'=> 'Maps\ServiceParam',
	);

	$wgParamDefinitions['mapslocation'] = array(
		'string-parser' => 'Maps\LocationParser',
	);

	$wgParamDefinitions['mapsline'] = array(
		'string-parser' => 'Maps\LineParser',
	);

	$wgParamDefinitions['mapscircle'] = array(
		'string-parser' => 'Maps\CircleParser',
	);

	$wgParamDefinitions['mapsrectangle'] = array(
		'string-parser' => 'Maps\RectangleParser',
	);

	$wgParamDefinitions['mapspolygon'] = array(
		'string-parser' => 'Maps\PolygonParser',
	);

	$wgParamDefinitions['distance'] = array(
		'string-parser' => 'Maps\DistanceParser',
	);

	$wgParamDefinitions['wmsoverlay'] = array(
		'string-parser' => 'Maps\WmsOverlayParser',
	);
} );
